decompy/RL/Model/Model_template.py

```diff
@@ -38,19 +38,61 @@ def get_past_info(self, summary_of_current_state, possible_actions):
             """
             pass
 
+        def get_decision_by_hash(self, hashed_llvm):
+            """
+            gets the decision by hashed_llvm summary.
+            :param hashed_llvm: the hashed_llvm summary
+            :return: a decision or None if not found.
+            :rtype: Decision
+            """
+            return self.decision_history[hashed_llvm]
+
+        def get_all_decisions_dictionary(self, hashed_llvm):
+            """
+            gets the decision by hashed_llvm summary.
+            :param hashed_llvm: the hashed_llvm summary
+            :return: a decision or None if not found.
+            :rtype: Decision
+            """
+            dict_list = []
+            # loop through all keys, add them to a new list
+            for key, value in self.decision_history:
+                if type(value) is list:
+                    dict_list.extend(value)
+                else:
+                    dict_list.append(value)
+            return self.decision_history[hashed_llvm]
+
+        def get_all_decisions_list(self):
+            """
+            gets all decisions and puts them into a list.
+            :return: a decision list
+            :rtype: list<Decision>
+            """
+            return (for _ in self.decision_history.iteritems())
+
         def update_model(self, decision):
             """
             Updates the model based off a decision.
             :param decision: the decision to update the model
             :type: Decision
             :return: nothing
             """
-            pass
+            # get hash representation from llvm summary
+            hashed = hash(decision.summary)
```

Comment:
Why are you hashing the summary manually? Dictionaries work by hashing the key object in python. That means you hash it, put that hash in, then that hash is hashed. It adds extra work for the programmer to put the hash in and it is less efficient.